lib/grant-types/authorization-code-grant-type.js

```diff
@@ -0,0 +1,170 @@
+
+/**
+ * Module dependencies.
+ */
+
+var AbstractGrantType = require('./abstract-grant-type');
+var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidGrantError = require('../errors/invalid-grant-error');
+var InvalidRequestError = require('../errors/invalid-request-error');
+var Promise = require('bluebird');
+var ServerError = require('../errors/server-error');
+var is = require('../validator/is');
+var util = require('util');
+
+/**
+ * Constructor.
+ */
+
+function AuthorizationCodeGrantType(options) {
+  options = options || {};
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.model.getAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getAuthorizationCode()`');
+  }
+
+  if (!options.model.revokeAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `revokeAuthorizationCode()`');
+  }
+
+  if (!options.model.saveToken) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveToken()`');
+  }
+
+  AbstractGrantType.call(this, options);
+}
+
+/**
+ * Inherit prototype.
+ */
+
+util.inherits(AuthorizationCodeGrantType, AbstractGrantType);
+
+/**
+ * Handle authorization code grant.
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+ */
+
+AuthorizationCodeGrantType.prototype.handle = function(request, client) {
```

Comment:
The name we choose will have to fit the following handlers:

```
AuthorizationCodeGrantType
ClientCredentialsType
PasswordGrantType
RefreshTokenGrantType
```

Since only one of them handles authorization, I'd prefer to avoid the `validateAndAuthorize` variants and would be fine with `handleGrant`.

Hmm, how about `handleRequest()` since the signature for all handlers is: `function(request, client)`?